indsl/filter/hilbert_huang_transform.py

```diff
@@ -0,0 +1,718 @@
+    indices = np.digitize(data, boundaries) - 1
+
+    # Assign sentinel value to out-of-bound observations
+    indices[out_of_bounds] = DROP_SENTINAL
+
+    return indices
+
+
+def compute_hilbert_huang_2d_spectrum(values_X, values_Z, x_boundaries):
+    """Calculate a 2D Hilbert-Huang power distribution in sparse format.
+
+    This utility function generates a sparse matrix that captures a two-dimensional
+    power distribution. It's typically not intended for direct end-user interaction.
+
+    Parameters
+    ----------
+    values_X : ndarray
+        A 2D array with values for the primary dimension, typically [samples x imfs]
+    values_Z : ndarray
+        A 2D array of amplitude or power corresponding to values_X
+    x_boundaries : ndarray
+        A vector defining bin boundaries for values_X
+
+    Returns:
+    -------
+    sparse_spectrum
+        A sparse representation of the 2D power distribution.
+
+    """
+    # Ensure inputs are verified in calling functions.
+    # Determine bin indices for primary dimension
+    primary_indices = bin_indices(values_X, x_boundaries)
+
+    # Establish indices for the temporal dimension and adjust shape to fit
+    original_indices = np.arange(primary_indices.shape[0])[:, np.newaxis]
+    time_indices = np.broadcast_to(original_indices, primary_indices.shape)
+
+    # Identify indices for the IMF dimension and adjust shape accordingly
+    original_imf_indices = np.arange(values_X.shape[1])[np.newaxis, :]
+    imf_indices = np.broadcast_to(original_imf_indices, primary_indices.shape)
+
+    # Generate COO coordinate matrix in a vectorized fashion
+    coordinate_matrix = np.c_[primary_indices.flatten(), time_indices.flatten(), imf_indices.flatten()].T
+
+    # Compute exclusions based on coordinates
+    exclusions = np.any(coordinate_matrix == DROP_SENTINAL, axis=0)
+
+    # Reshape exclusions based on its length
+    reshaped_exclusions = exclusions.reshape(values_Z.shape)
+
+    # Mask the Z-values based on the exclusions matrix
+    values_Z_masked = np.ma.masked_array(values_Z, mask=reshaped_exclusions)
+    values_Z_masked = values_Z_masked.squeeze()
+
+    # Flatten the masked array to remove masked values
+    pointwise_exclusions = reshaped_exclusions.squeeze()
+
+    # Determine the valid points based on pointwise exclusions (i.e. any row-wise exclusions)
+    valid_points = ~pointwise_exclusions
+
+    # Use this to index the flattened Z-values and the coordinates
+    adjusted_coordinate_matrix = coordinate_matrix[:, valid_points]
+
+    # Determine the final matrix dimensions
+    spectrum_shape = (x_boundaries.shape[0] - 1, primary_indices.shape[0], primary_indices.shape[1])
+
+    # Construct the sparse spectrum representation
+    from sparse import COO
+
+    sparse_spectrum = COO(adjusted_coordinate_matrix, values_Z_masked, shape=spectrum_shape)
+
+    return sparse_spectrum
+
+
+def validate_dimensionality(arrays_to_validate, variable_names, calling_function):
```

Comment:
1. remove logging
2. remove variable names and calling_function
4. Make it internal function
5. Adjust and simplify docs: remove return and params as it is internal function
6. Add returning type to the function signature
7. Maybe make this function to validate one array at the time. And adjust the use of the function.